[SPARK-49505][SQL] Create new SQL functions "randstr" and "uniform" to generate random strings or numbers within ranges

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR introduces two new SQL functions "randstr" and "uniform" to generate random strings or numbers within ranges.

* The "randstr" function returns a string of the specified length whose characters are chosen uniformly at random from the following pool of characters: 0-9, a-z, A-Z. The random seed is optional. The string length must be a constant two-byte or four-byte integer (SMALLINT or INT, respectively).
* The "uniform" function returns a random value with independent and identically distributed  values with the specified range of numbers. The random seed is optional. The provided numbers specifying the minimum and maximum values of the range must be constant. If both of these numbers are integers, then the result will also be an integer. Otherwise if one or both of these are floating-point numbers, then the result will also be a floating-point number.

For example:

```
SELECT randstr(5);
> ceV0P

SELECT randstr(10, 0) FROM VALUES (0), (1), (2) tab(col);
> ceV0PXaR2I
  fYxVfArnv7
  iSIv0VT2XL

SELECT uniform(10, 20.0F);
> 17.604954

SELECT uniform(10, 20, 0) FROM VALUES (0), (1), (2) tab(col);
> 15
  16
  17
```

### Why are the changes needed?

This improves the SQL functionality of Apache Spark and improves its parity with other systems:
* https://clickhouse.com/docs/en/sql-reference/functions/random-functions#randuniform
* https://docs.snowflake.com/en/sql-reference/functions/uniform
* https://www.microfocus.com/documentation/silk-test/21.0.2/en/silktestclassic-help-en/STCLASSIC-8BFE8661-RANDSTRFUNCTION-REF.html
* https://docs.snowflake.com/en/sql-reference/functions/randstr

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds golden file based test coverage.

### Was this patch authored or co-authored using generative AI tooling?

Not this time.

dtenedor committed:
commit

commit

uniform expression

commit

commit

commit

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -384,7 +384,9 @@ object FunctionRegistry {
     expression[Rand]("rand"),
     expression[Rand]("random", true, Some("3.0.0")),
     expression[Randn]("randn"),
+    expression[RandStr]("randstr"),
```
Like the previous work, I can also add DataFrame API support for these new functions as well in a different PR (to separate concerns).

dtenedor commented:
cc @HyukjinKwon @MaxGekk here are a couple more :)

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -384,7 +384,9 @@ object FunctionRegistry {
     expression[Randn]("randn"),
+    expression[RandStr]("randstr"),
     expression[Stack]("stack"),
+    expression[Uniform]("uniform"),
```
Should also add it into Scala/Python functions in both Spark Connect and Classic. Would be great if we can file JIRAs to make sure we don't forget

dtenedor replied:
Sounds good I created this Jira in https://issues.apache.org/jira/browse/SPARK-49552

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +187,211 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+       abc
+  """,
+  since = "4.0.0",
+  group = "math_funcs")
```
How about `string_funcs`?

dtenedor replied:
Sounds good, done.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +187,211 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+
+@ExpressionDescription(
+  usage = """
+    _FUNC_(length, seed) - Returns a string of the specified length whose characters are chosen
```
Mark `seed` as optional:
```suggestion
    _FUNC_(length[, seed]) - Returns a string of the specified length whose characters are chosen
```

dtenedor replied:
Sounds good, done.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +187,211 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+          bytes.update(i, ('A' + num - 20).toByte)
+      }
+    }
+    val result: UTF8String = UTF8String.fromBytes(bytes.toArray)
```
```suggestion
    val result: UTF8String = UTF8String.fromBytes(bytes)
```

dtenedor replied:
Sounds good, done.

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
fix test

dtenedor committed:
fix test

dtenedor committed:
fix test

dtenedor committed:
commit

commit

dtenedor committed:
commit

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/ExpressionsSchemaSuite.scala:
```diff
@@ -118,11 +118,11 @@ class ExpressionsSchemaSuite extends QueryTest with SharedSparkSession {
         example.split("  > ").tail.filterNot(_.trim.startsWith("SET")).take(1).foreach {
           case _ if funcName == "from_avro" || funcName == "to_avro" ||
-            funcName == "from_protobuf" || funcName == "to_protobuf" =>
+            funcName == "from_protobuf" || funcName == "to_protobuf" || funcName == "uniform" =>
```
Could you elaborate the changes, please. As I can see, the file `sql-expression-schema.md` contains some random like expr/functions, see:
```
| org.apache.spark.sql.catalyst.expressions.Rand | rand | SELECT rand() | struct<rand():double> |
| org.apache.spark.sql.catalyst.expressions.Rand | random | SELECT random() | struct<rand():double> |
```
Why do you want to exclude `uniform` from it?

dtenedor replied:
I updated this so `uniform` is no longer excluded.

MaxGekk commented on sql/core/src/test/resources/sql-functions/sql-expression-schema.md:
```diff
@@ -265,6 +265,7 @@
 | org.apache.spark.sql.catalyst.expressions.RaiseErrorExpressionBuilder | raise_error | SELECT raise_error('custom error message') | struct<raise_error(USER_RAISED_EXCEPTION, map(errorMessage, custom error message)):void> |
 | org.apache.spark.sql.catalyst.expressions.Rand | rand | SELECT rand() | struct<rand():double> |
 | org.apache.spark.sql.catalyst.expressions.Rand | random | SELECT random() | struct<rand():double> |
+| org.apache.spark.sql.catalyst.expressions.RandStr | randstr | SELECT randstr(3, 0) | struct<randstr(3, 0):string> |
```
Why only `RandStr`? Where is `Uniform`?

dtenedor replied:
I updated this so `uniform` is no longer excluded.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +189,215 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+          errorSubClass = "UNEXPECTED_INPUT_TYPE",
+          messageParameters = Map(
+            "paramIndex" -> ordinalNumber(index),
+            "requiredType" -> "constant value of integer or floating-point",
```
Precisely speaking, you require a foldable expr not just a constant.

dtenedor replied:
Good point, updated the error messages accordingly.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +189,215 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+    Seq(min, max, seed).zipWithIndex.foreach { case (expr: Expression, index: Int) =>
+      if (!valid(expr)) {
+        result = DataTypeMismatch(
+          errorSubClass = "UNEXPECTED_INPUT_TYPE",
```
We special error condition for non-foldable args: `NON_FOLDABLE_INPUT`. Can you return it in the case of non-foldable expr, please.

dtenedor replied:
Sounds good, done.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +189,215 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+    Uniform(newFirst, newSecond, newThird)
+
+  override def toString: String = prettyName + truncatedString(
+    Seq(min, max), "(", ", ", ")", SQLConf.get.maxToStringFields)
```
Do you exclude `seed` intentionally?

dtenedor replied:
I reverted this method override now.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +189,215 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+  }
+
+  private def integer(t: DataType): Boolean = t match {
+    case _: ShortType | _: IntegerType | _: LongType => true
```
You could check either it is `IntegralType` or not if you would support `ByteType`.

dtenedor replied:
Good question; we specifically want to exclude `ByteType` here, thus the explicit list of integer types.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks @MaxGekk for your reviews, responded to your comments, please take another look.

dtenedor committed:
fix function description

dtenedor committed:
fix test

dtenedor committed:
fix test

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +187,239 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+  group = "math_funcs")
+object UniformExpressionBuilder extends ExpressionBuilder {
+  override def build(funcName: String, expressions: Seq[Expression]): Expression = {
+    val numArgs = expressions.length
```
nit: could you move it into the default case

dtenedor replied:
👍 this is done

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +187,239 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+}
+
+object Uniform {
+  lazy val random = new Random()
```
Why don't you follow the existing approach for the random expression, and don't extend `RDG`?

dtenedor replied:
Good question, `RDG` `extends UnaryExpression`, whereas this new function accepts two arguments.

dtenedor replied:
Update on this: I was able to split the `UnaryExpression` part of `RDG` into a separate class and now `Uniform` can reuse the remainder of `RDG` to simplify this.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +187,239 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+case class Uniform(min: Expression, max: Expression)
+  extends RuntimeReplaceable with BinaryLike[Expression] with ExpressionWithRandomSeed {
+
+  private var seed: Expression = Literal(Uniform.random.nextLong(), LongType)
```
hmm, why do you need `var`. Could you follow the existing approach, and define a case class w/ 3 parameters + constructors with 2 params.

dtenedor replied:
Please see above reply.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +187,239 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+  """,
+  since = "4.0.0",
+  group = "math_funcs")
+object UniformExpressionBuilder extends ExpressionBuilder {
```
Why do you need an expression builder? Can't you just define an expression w/ a few constructors? Please, explain this.

dtenedor replied:
I originally tried that, but found that putting the random seed expression as a case class field made it get printed in all sorts of places (even if I overrode `sql` and/or `toString`). The value is non-deterministic which would cause flakiness. So I moved the seed to a private field of the class instead, but this precludes setting it in extra constructors, thus arriving at this approach.

dtenedor replied:
@MaxGekk update: alright I was able to remove both the `ExpressionBuilder` and the `var` and just add the random seed as a class field, per your suggestion.

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +189,218 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+    result
+  }
+
+  override def evalInternal(input: InternalRow): Any = {
```
This non-codegen implementation is not covered by any tests. I replaced it by:
```
  override def evalInternal(input: InternalRow): Any = {
    throw new java.lang.IllegalArgumentException("evalInternal")
  }
```
and run `random.sql` which checked the codegen impl but not this one.

dtenedor replied:
Busted :) I added some separate testing in `RandomSuite.scala` for the non-codegen case.

MaxGekk reviewed (changes requested):
Test the non-codegen implementation.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks again @MaxGekk for the thorough reviews
they are helping the new functions cover all corner cases as carefully as possible.

MaxGekk reviewed (commented):
LGTM in general. @dtenedor Could you fix the test failure:
```
[info] - SPARK-49505: Test the RANDSTR and UNIFORM SQL functions without codegen *** FAILED *** (1 millisecond)
[info]   Exception evaluating randstr(10, 0) (ExpressionEvalHelper.scala:257)
...
[info]   Cause: java.lang.ClassCastException: class java.lang.Long cannot be cast to class java.lang.Integer (java.lang.Long and java.lang.Integer are in module java.base of loader 'bootstrap')
[info]   at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:99)
[info]   at org.apache.spark.sql.catalyst.expressions.RandStr.evalInternal(randomExpressions.scala:377)
```

MaxGekk reviewed (commented):
> and improves its parity with other systems.

@dtenedor Could you provide a few links to the systems that you are going to reach the parity.

dtenedor commented:
> @dtenedor Could you provide a few links to the systems that you are going to reach the parity.
* https://clickhouse.com/docs/en/sql-reference/functions/random-functions#randuniform
* https://docs.snowflake.com/en/sql-reference/functions/uniform
* https://www.microfocus.com/documentation/silk-test/21.0.2/en/silktestclassic-help-en/STCLASSIC-8BFE8661-RANDSTRFUNCTION-REF.html
* https://docs.snowflake.com/en/sql-reference/functions/randstr

dtenedor committed:
fix RandomSuite

dtenedor commented:
> LGTM in general. @dtenedor Could you fix the test failure

👍 this is done

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +188,237 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+  }
+
+  override def evalInternal(input: InternalRow): Any = {
+    val numChars: Int = length.eval(input) match {
+      case i: Int => i
+      case n: Long => n.toInt
+      case s: Short => s.toInt
+    }
```
nit:
```suggestion
    val numChars = length.eval(input).asInstanceOf[Number].intValue()
```

dtenedor replied:
thanks, this is better

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -181,3 +188,237 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+        case _ if num < 10 =>
+          bytes.update(i, ('0' + num).toByte)
+        case _ if num < 20 =>
+          bytes.update(i, ('a' + num - 10).toByte)
```
And how your algorithm returns 'z', for instance?
```
'a' = 97, 'z' = 122
122 = 97 + num - 10 -> num = 35
```
the first case:
```
'0' = 48
122 = 48 + num -> num = 74
```
the last one:
```
'A' = 65
122 = 65 + num - 20 -> num = 77
```

To get 'z', num must be greater than `(rng.nextInt() % 30).abs`.

Your expression description says:
```
characters are chosen uniformly at random from the following pool of characters: 0-9, a-z, A-Z
```

There is some contradiction, isn't it?

dtenedor replied:
You are correct, this was a bug in the algorithm. The random number range should actually be between 0 and 61, inclusive. Then the range [0, 9] corresponds to digits 0-9, [10-35] corresponds to characters a-z, and [36-61] corresponds to A-Z. I updated the algorithm to fix it and added a comment here.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks again @MaxGekk for your reviews

MaxGekk reviewed (approved):
Waiting for CI.

@dtenedor Please, update PR description (examples and so on) according to your recent changes.

MaxGekk commented:
@dtenedor Could you fix the test failure, please. It seems it is related to your changes.
```
[info]   Cause: org.scalatest.exceptions.TestFailedException: Function 'randstr', Expression class 'org.apache.spark.sql.catalyst.expressions.RandStr' "[ceV]" did not equal "[8i7]"
```

dtenedor committed:
fix test

MaxGekk commented:
+1, LGTM. Merging to master.
Thank you, @dtenedor and @HyukjinKwon for review.
